### PR TITLE
doc: add swagger support

### PR DESCRIPTION
**Pull request by loxp**
### What problem does this PR solve? 
UCP #2169 

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

 - Has configuration change

Side effects

Related changes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -131,6 +140,10 @@ lint:
 	@echo "linting"
 	CGO_ENABLED=0 revive -formatter friendly -config revive.toml $$($(PACKAGES))
 
+swagger_spec:
+	go mod vendor
+	swag init --parseVendor -generalInfo server/api/router.go --exclude vendor/github.com/pingcap-incubator/tidb-dashboard --output docs/swagger
```
Comment (nolouch) [inline Makefile]: Do we need to generate for `tidb-dashboard`? cc @HunDunDM

Comment (nolouch) [reply Makefile]: and we need  to help users to install `swag` in scripts.

Comment (loxp) [reply Makefile]: > Thanks @loxp , Sorry too late for review. Do you test the swagger API? After build `pd-server` with `SWAGGER=1 make`, and swagger API not works.
> 
> ```
> ❯ curl http://127.0.0.1:2379/pd/api/v1/swagger
> Not Found
> ```
> 
> and also after I `curl` the API, the pd-server cannot be stoped by `Ctrl+C`.

I add a specific router for `/pd/api/v1/swagger` in the latest commit, PTAL.

PD is successfully stoped by `Ctrl+C` on my macOS Catalina. Is it different from your environment?

Comment (loxp) [reply Makefile]: > and we need to help users to install `swag` in scripts.

It is added in tools.go so can be installed in `make install-tools`.

File: server/api/router.go [hunk 2]
```diff
@@ -194,6 +204,11 @@ func createRouter(ctx context.Context, prefix string, svr *server.Server) (*mux.
 	apiRouter.Handle("/debug/pprof/block", pprof.Handler("block"))
 	apiRouter.Handle("/debug/pprof/goroutine", pprof.Handler("goroutine"))
 
+	// swagger docs
+	swaggerHandler := swaggerserver.Handler()
+	apiRouter.PathPrefix("/swagger/").Handler(swaggerHandler)
```
Comment (HunDunDM) [inline server/api/router.go]: Is it better to add a `HandlerBuilder` to `main.go` and access it directly via `/swagger`?

**Comment by HunDunDM** [comment]
I think you need to create a new empty file `placeholder.go` in the `docs` directory to prevent errors.
In addition, there are some `FIXME` in the comments. How to deal with this in the future?

File: go.mod [hunk 3]
```diff
@@ -47,16 +51,17 @@ require (
 	github.com/sirupsen/logrus v1.2.0
 	github.com/spf13/cobra v0.0.3
 	github.com/spf13/pflag v1.0.1
+	github.com/swaggo/http-swagger v0.0.0-20200103000832-0e9263c4b516
+	github.com/swaggo/swag v1.6.6-0.20200323071853-8e21f4cefeea
```
[the remark this answers is not in the document]
Comment (loxp) [reply go.mod]: v1.6.5 does not have the `-exclude` flag which is need to exclude the tidb-dashboard docs.

**Comment by loxp** [comment]
> In addition, there are some `FIXME` in the comments. How to deal with this in the future?

To fix the FIXMEs, I think it needs to use struct to unmarshal the parameters instead of map. Maybe it is better to fix later. What do you think?

File: go.mod [hunk 4]
```diff
@@ -4,6 +4,7 @@ go 1.13
 
 require (
 	github.com/BurntSushi/toml v0.3.1
+	github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751
```
[the remark this answers is not in the document]
Comment (loxp) [reply go.mod]: It was used in `docs/swagger/doc.go`. Now I remove it.

File: go.mod (hunk 3 above)
[the remark this answers is not in the document]
Comment (HunDunDM) [reply go.mod]: After discussing with @nolouch , we think it doesn't need to exclude the tidb-dashboard docs.

File: server/api/router.go (hunk 2 above)
Comment (loxp) [reply server/api/router.go]: `/pd/swagger` or `/swagger`?

File: Makefile [hunk 5]
```diff
@@ -131,6 +140,10 @@ lint:
 	@echo "linting"
 	CGO_ENABLED=0 revive -formatter friendly -config revive.toml $$($(PACKAGES))
 
+swagger_spec:
```
Comment (HunDunDM) [inline Makefile]: Need to add the `install-tools` target.

File: .gitignore [hunk 6]
```diff
@@ -13,3 +13,4 @@ default*
 .vscode/
 /.dashboard_asset_cache
 /.tools/bin
+docs/swagger*
```
Comment (HunDunDM) [inline .gitignore]: Create a new file `placeholder.go` in the directory and specify other files to ignore it.

File: docs/swagger/docs.go [hunk 7]
```diff
@@ -0,0 +1 @@
+package swagger
```
Comment (HunDunDM) [inline docs/swagger/docs.go]: Rename it to `placeholder.go` and don't ignore it in `.gitignore`.

File: Makefile [hunk 8]
```diff
@@ -56,13 +62,16 @@ build: pd-server pd-ctl
 tools: pd-tso-bench pd-recover pd-analysis pd-heartbeat-bench
 pd-server: export GO111MODULE=on
 pd-server:
+ifeq ($(SWAGGER), 1)
```
[the remark this answers is not in the document]
Comment (nolouch) [reply Makefile]: agree

**Pull request merged**